libpod/container_top_linux.go

```diff
@@ -113,60 +118,146 @@ func (c *Container) GetContainerPidInformation(descriptors []string) ([]string,
+			fields := strings.Fields(line)
+			if len(fields) > 3 {
+				// open the shared lib on the host as it will most likely not be in the container
+				logrus.Debugf("Open shared library for ps: %s", fields[2])
+				fd, err := unix.Open(fields[2], unix.O_PATH, 0)
+				if err == nil {
+					defer unix.Close(fd)
+					preload = append(preload, fmt.Sprintf("/proc/self/fd/%d", fd))
+				}
+			} else if len(fields) == 2 {
+				if path.IsAbs(fields[0]) {
+					// this should be the dynamic linker
+					logrus.Debugf("Using linker for ps: %s", fields[0])
+					linkFD, err := unix.Open(fields[0], unix.O_PATH|unix.O_CLOEXEC, 0)
+					if err != nil {
+						return nil, err
+					}
+					defer unix.Close(linkFD)
+					linkerPath = fmt.Sprintf("/proc/self/fd/%d", linkFD)
+				}
+			}
 		}
+		// Ok, set linker args. First overwrite argv[0] because busybox for example needs it to know
+		// which program to execute as everything is in one binary and they need to proper name.
+		// Second now preload all linked shared libs. This is to prevent the executable from loading
+		// any libs in the container and thus very likely failing.
+		args = append([]string{linkerPath, "--argv0", "ps", "--preload", strings.Join(preload, " ")}, args...)
 	}
 
-	return stdout, nil
+	pid := c.state.PID
+	errChan := make(chan error)
+	go func() {
+		defer close(errChan)
+
+		// DO NOT UNLOCK THIS THREAD!!!
+		// We are joining a different pid and mount ns, go must destroy the
+		// thread when we are done and not reuse it.
+		runtime.LockOSThread()
+
+		// join the mount namespace of pid
+		mntFD, err := os.Open(fmt.Sprintf("/proc/%d/ns/mnt", pid))
+		if err != nil {
+			errChan <- err
+			return
+		}
+		defer mntFD.Close()
+
+		// join the pid namespace of pid
+		pidFD, err := os.Open(fmt.Sprintf("/proc/%d/ns/pid", pid))
+		if err != nil {
+			errChan <- err
+			return
+		}
+		defer pidFD.Close()
+
+		// create a new mountns on the current thread
+		if err = unix.Unshare(unix.CLONE_NEWNS); err != nil {
+			errChan <- fmt.Errorf("unshare NEWNS: %w", err)
+			return
+		}
+		if err := unix.Setns(int(mntFD.Fd()), unix.CLONE_NEWNS); err != nil {
+			errChan <- fmt.Errorf("setns NEWNS: %w", err)
+			return
+		}
+
+		if err := unix.Setns(int(pidFD.Fd()), unix.CLONE_NEWPID); err != nil {
+			errChan <- fmt.Errorf("setns NEWPID: %w", err)
+			return
+		}
+
+		logrus.Debugf("Executing ps in the containers mnt+pid namespace, final command: %v", args)
+		var errBuf bytes.Buffer
+		path := args[0]
+		args[0] = "ps"
+		cmd := exec.Cmd{
+			Path:   path,
+			Args:   args,
+			Stdout: wPipe,
+			Stderr: &errBuf,
+		}
+
+		err = cmd.Run()
+		if err != nil {
+			exitError := &exec.ExitError{}
+			if errors.As(err, &exitError) && errBuf.Len() > 0 {
+				// when error printed on stderr include it in error
+				err = fmt.Errorf("ps failed with exit code %d: %s", exitError.ExitCode(), errBuf.String())
+			} else {
+				err = fmt.Errorf("could not execute ps in the container: %w", err)
```

Comment:
```suggestion
				err = fmt.Errorf("could not execute host ps(1) in the container: %w", err)
```

To avoid ambiguity with what we did prior to this PR.